### PR TITLE
`Integrated code lifecycle`: Structure buildlogs folder

### PR DESCRIPTION
**Pull request by BBesrour**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Currently, all the build log files are saved under the same folder. This raises an issue when admins want to access a build log file manually where the folder takes a long time to load due to the extremely large number of files 

### Description

This PR introduces a folder structure to the build logs:
- build-logs
- - Course-short-name
- - - Exercise-short-name
- - - - buildlog file

Old build logs files will stay in that parent folder for now and will be automatically gradually deleted.

### Steps for Testing

Prerequisites (Test servers):
- Before checking out this branch, make a few submissions from develop
- 1 Instructor

1. Create a programming exercise and submit a few attempts
2. Go to Build Overview
3. Check that the old and new build log files are accessible
4. Go to exercise participation, check the submissions for a participation (ideally a new submission and an old one), no errors

Prerequisites (Optional - Locally):
- Before checking out this branch, make a few submissions from develop
- 1 Instructor

1. Check that the build logs are saved correctly (they should be under a directory with path './build-logs' relative to the artemis repo)

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **New Features**
	- Enhanced build log handling with improved organization and access based on programming exercises.
	- Introduced a method to check for the existence of log files in both exercise-specific and parent directories.
	- Added functionality to process results with a focus on programming exercise participation.
	- New methods in repositories for robust retrieval and error handling of build jobs and programming exercises.

- **Bug Fixes**
	- Implemented a fallback mechanism for backward compatibility with existing logs.

- **Documentation**
	- Updated method signatures to reflect new parameters and functionality related to participation.

- **Tests**
	- Improved test cases to validate new functionality regarding build logs and participation context.

### CODE REVIEW

File: src/main/java/de/tum/cit/aet/artemis/programming/service/BuildLogEntryService.java [hunk 1]
```diff
@@ -359,25 +407,86 @@ public void deleteOldBuildLogsFiles() {
...
+            deleteRecursively(buildLogsPath);
         }
         catch (IOException e) {
             log.error("Error occurred while trying to delete old build log files", e);
         }
     }
 
-    public boolean buildJobHasLogFile(String buildJobId) {
-        Path logPath = buildLogsPath.resolve(buildJobId + ".log");
+    private void deleteRecursively(Path path) throws IOException {
```
Comment (b-fein) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/BuildLogEntryService.java:421]: Does the library method in Spring already implement this sufficiently?

https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/FileSystemUtils.html#deleteRecursively(java.nio.file.Path)

**Review by SimonEntholzer — DISMISSED** [review]
Work fine locally :+1: 
- creates logs in new folder structure
- can view old and new logs in the build overview

File: src/main/java/de/tum/cit/aet/artemis/programming/service/BuildLogEntryService.java (hunk 1 above)
Comment (BBesrour) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/BuildLogEntryService.java:421]: I didn't about that utility function! But yes, we can't use due the expired check

Comment (pzdr7) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/BuildLogEntryService.java:421]: The method name `deleteRecursively()` seems a bit generic, given that it also checks if the files are expired. Could you be more specific here? e.g. deleteExpiredBuildLogFilesRecursively or something similar?

**Review by pzdr7 — CHANGES_REQUESTED** [review]

**Review by pzdr7 — APPROVED** [review]
Thanks for incorporating the changes 👍

**Review by coolchock — APPROVED** [review]
code LGTM

**Review by GODrums — APPROVED** [review]
Tested on testing server in review session, LGTM.

**Review by cremertim — APPROVED** [review]
Tested in testing session yesterday, also retested locally today. Changes lgtm

**Pull request merged**